### PR TITLE
Refactor and fix multi-facet selection.

### PR DESCRIPTION
**Pull request by EreMaijala**
- Simplifies the code
- Fixes filter value handling when the value in URL is not enclosed in quotes
- Uses more CSS to hide and show elements
- Reduces JS manipulation of DOM
- Fixes display of selected facets in facet lightbox
- Improves the visuals of facet lightbox
- Doesn't cause errors if sidebar is not found
- Adds multiselection to standalone facet list page
- Fixes multi-selection after loading more facets in lightbox
- Fixes hiding of Set buttons loaded after activating multi-facet selection
- Fixes autocomplete results showing under multi-facet selection checkbox

Fixes VUFIND-1719 and VUFIND-1720.

TODO:
- [x] Add Mink tests
- [x] Mark VUFIND-1719 and VUFIND-1720 fixed when committed.

### CODE REVIEW

**Comment by EreMaijala** [comment]
@sturkel89 This changes a lot of the multi-facet selection code, so I'd appreciate some testing with different themes and including the facet lightbox. :)

**Review by sturkel89 — CHANGES_REQUESTED** [review]
Thanks for working on this, Ere!

As far as I can tell, multiple filters are working properly from the main screen in all five themes, both without any customizations and also when I enable OR and EXCLUDE facets.

However, _multiple filters are NOT working properly in lightbox any of the themes_. When I open any of the lightboxes, the **Submit multiple filters** checkbox is visible, but when I check it there's no **Apply filters** button as there is in the main interface.

Therefore, when I check the box and then click one of the filters, it's applied immediately and I'm popped back to the results screen. There's no opportunity to actually select more than one filter. 

I think we went through this when @rominail introduced this feature in PR 3761. I believe he fixed it then, so hopefully it's an easy fix now.

---------

I'd also like to comment on the styling of the **Select multiple filters** item and **Apply filters** button at the top of the facet area. There are variations between themes in terms of how the new elements relate to the size and/or boldness of the other elements. I've documented that below. 

My recommendation is that the **Select multiple filters** text be BOLDED in every case and match the size of the facet items, and that the **Apply filters** text also match the size of the facet items but NOT be bolded. 

### Bootstrap3: 
image

Select element is 14px and bold
Apply button text is 14px and normal
Facet headers are 14px and bold
Facet items are 14px and normal.

### Bootstrap5: 
image

Select element is 14px and normal
Apply button text is 14px and normal
Facet headers are 14px and bold
Facet items are 14px and normal.

### Bootprint3: 
image

Select element is 13px and bold
Apply button text is 14px and bold
Facet headers are 14px and bold
Facet items are 13px and normal

### Sandal: 
image

Select element is 14px and bold
Apply button text is 16px and normal
Facet headers are 16px and bold
Facet items are 14px and normal

### Sandal5:
image

Select element is 14.4px and normal
Apply button text is 16px and normal
Facet headers are 16px and bold
Facet items are 14.4px and normal

File: module/VuFind/src/VuFind/Search/UrlQueryHelper.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/UrlQueryHelper.php]: Have you tested whether this works as expected with complicated multi-part checkbox filters (e.g. `format:Book OR title:test`)? If not, maybe we should add a test case for that scenario....

File: themes/bootstrap5/templates/Recommend/SideFacets/single-facet.phtml
Comment (demiankatz) [inline themes/bootstrap5/templates/Recommend/SideFacets/single-facet.phtml:60]: Does `$hasSubLinks` really no longer need to exist? If so, you can probably stop defining it above.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@EreMaijala, I should probably wait until you've addressed @sturkel89's review before giving this a really close look, but I skimmed through just to get a feel for the scope and spotted a couple of questions along the way.

Also, should we list in this PR which JIRA ticket(s) are addressed by this work, if indeed some are addressed? Or is this all separate from those two open multi-facet selection tickets?

**Comment by demiankatz** [comment]
Also, should this target the release-10.1 branch since it's fixing broken functionality? Or is the concern that the cure is worse than the disease here?

**Comment by EreMaijala** [comment]
@demiankatz Good point about the release-10.1 branch. I'll try to rebase. It will need a bit of adjustment because of the changes with checkbox facet counts.

File: module/VuFind/src/VuFind/Search/UrlQueryHelper.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Search/UrlQueryHelper.php]: Oh no. Of course it doesn't, but with the normalization in JS we shouldn't need this change, so I'll just revert it.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/Search/UrlQueryHelper.php]: But then the JS side may also need a bit of adjustment to handle this sort properly.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/Search/UrlQueryHelper.php]: Maybe I misunderstood. If the complex filters are a result of CustomFilters, then the JS side is unaffected.

**Comment by EreMaijala** [comment]
@sturkel89 Oops, sorry. I messed the facet lightbox up during final cleanup. Should be fixed now. I also made some tweaks to boldness and font size according to your wishes.

File: themes/bootstrap5/templates/Recommend/SideFacets/single-facet.phtml
Comment (EreMaijala) [reply themes/bootstrap5/templates/Recommend/SideFacets/single-facet.phtml:60]: Yes, removed.

**Comment by EreMaijala** [comment]
Now based on release-10.1. I tried to ensure good forward-compatibility, but the removed checkboxFilter class remains in 10.1 for back-compatibility and will probably cause a merge conflict.

**Comment by GrothTobias** [comment]
Just a question: Is it possible to enable multi-facet selection by default? 
I thought the easiest way is to just add a `checked` in the template but that does not work and creates unwanted behaviour - the select button is displayed when unchecking the checkbox and hidden when checking the checkbox.

**Comment by EreMaijala** [comment]
@GrothTobias Currently, no, but I'm planning to add that as an option.

**Comment by EreMaijala** [comment]
@demiankatz I've added some tests that verify that multi-facet selection works properly. At least the checkbox facet selection tests fail without the fixes. Facet lightbox doesn't have so good coverage yet, but it would require a bit more work than I can do right now. I tried to add the tests so that they don't conflict with the new ones already in dev branch, but we may want to see if they could be combined after the merge.

**Review by sturkel89 — CHANGES_REQUESTED** [review]
Lightbox is now working under certain circumstances.

When I opened a lightbox and left everything at the default, I could check the **Select multiple filters box** and the **Apply filters** button appeared. I could then check the boxes for some filter values and click the X to exclude some, and each of the excluded items got the orange outline. When I applied filters, all of my filters were correctly applied. So, things work in the default configuration.

When I opened a lightbox and either switched to Alphabetical sorting, or entered any text into the Filter box to see a shorter list of filter values, things didn't work as well. I tried this using the Library lightbox since it allows both alphabetical sorting and filtering by entering a string into a box.

- In some experiments, I immediately switched to Alphabetical before clicking **Select multiple filters**, and when I checked the box to Select multiple filters the **Apply filters** button did NOT appear! The same thing happened when I tried to limit by typing a letter into the Filter box to get a shorter list.

- In other experiments, I clicked **Select multiple filters** first, got the **Apply filters** button, and then switched to Alphabetical; in those cases, I could checkbox some values to OR (include) in my filtering, and they turned orange correctly. **However, when I tried to exclude some filters, I did not get any orange outline around the X**. Nevertheless, the filters were correctly applied. The same thing happened when I tried to limit by typing a letter into the Filter box to get a shorter list.

------------
In terms of the font styling for the elements, I can see that the bolding has been fixed across the board in all themes - looks good! However, it looks to me like the font size in the **Apply filters** button is larger than the font size of the **Select multiple filters** text in the following themes:

- bootprint3
- sandal

File: module/VuFind/src/VuFind/Search/UrlQueryHelper.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Search/UrlQueryHelper.php]: As discussed on Slack, I was confusing hidden filters and checkbox filters, so I think all is well. Closing out this thread now.

**Comment by demiankatz** [comment]
Thanks for all the progress here, @EreMaijala! I'll wait for you to respond to @sturkel89's latest reports before I take a final close review of everything... but if you need my help with anything in the meantime, please let me know!

**Comment by EreMaijala** [comment]
@sturkel89 Oops, I failed to initialize the alphabetical list properly. Should be fixed now. And styles tweaked as well.

**Comment by EreMaijala** [comment]
@demiankatz I also added a basic test for multiselection in facet lightbox. It's by no means complete, but at least it's a start.

**Review by sturkel89 — APPROVED** [review]
I checked **Select multiple filters** in each lightbox configuration: default, alphabetical, limited by typing something into the filter box, and THEN clicking the multifilter option. Everything is working correctly, including the persistent visible outlines around the X for excluded filters. Looks good!

Also, the font sizes and formatting are all fixed for the two new elements in every theme. Thanks, @EreMaijala!!

-------------
I noticed something new that is not a big deal at all: 

In the legacy themes, when you're in the lightbox and click in the filter box to limit the list of options, **the bottom part of the outlining of the text entry box isn't visible**. I think it might be overlaid by the **Select multiple filters** element – does that element invisibly stretch all the way across the lightbox? The affected themes are:

- bootstrap3
- bootprint3
- sandal

Here's an example of how that looks:

image

Bootstrap5 looks good, as does sandal5. Here's a screenshot of bootstrap5:

image

I'm going to approve because all of the functionality is present. If you can fix this styling issue as well, please go ahead!

**Comment by demiankatz** [comment]
Thanks, @sturkel89! Regarding the outlining issue, it would theoretically be nice to fix that, but if it proves non-trivial, I wouldn't worry too much since those legacy themes are all going away soon anyway.

File: module/VuFind/src/VuFindTest/Feature/SearchFacetFilterTrait.php
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Feature/SearchFacetFilterTrait.php]: Is this comment accurate? Seems like it takes any selector, so is "by data-title attribute" just an example?

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: This repeated logic might usefully be refactored to an `assertResultAndFilterCounts()` method. Not a must-have, but just a thought.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: Inconsistent capitalization:

```suggestion
    public static function checkboxFacetSelectionProvider(): array
```

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: ...and again here, so I guess it's consistently inconsistent. ;-)

```suggestion
     * @dataProvider checkboxFacetSelectionProvider
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
I'm still in the process of reviewing everything, but here is a partial review focused on tests.

**Comment by EreMaijala** [comment]
@sturkel89 Outlining should be fixed now!

File: module/VuFind/src/VuFindTest/Feature/SearchFacetFilterTrait.php
Comment (EreMaijala) [reply module/VuFind/src/VuFindTest/Feature/SearchFacetFilterTrait.php]: Fixed.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: Improved that and another one.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: Fixed.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php]: Fixed, now consistently consistent. :)

**Review by demiankatz — APPROVED** [review]
Looks good! Thanks, @EreMaijala and @sturkel89!

File: themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml
Comment (ThoWagen) [inline themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: @EreMaijala @demiankatz Was it intentional that the link does not fill the whole `li` element and the badge is not part of it anymore?

The behavior is now different to the account menu (badges for favorite lists) and it does not work that well with our local design.

Comment (demiankatz) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: @ThoWagen, are you seeing any visual differences in the core themes, or just in your local customizations? I don't believe there was any intent to change things visually. If the restructuring is having negative side effects, we should at very least add a changelog note about it. @EreMaijala can probably answer better than I can regarding intent/strategy since he did most of the work here. (He's also more likely to be around to discuss further this week since I'm currently on vacation... but I'll try to keep an eye on this in case my help is needed).

Comment (ThoWagen) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: In the core themes you can not select the facets by clicking on the badge or on the border of the list item anymore. I would expect that to work as a user but maybe only because I'm used to it.

We also change the background color of the list item on hovering in our local theme. But now the background color would change even if nothing can be clicked yet. That might be confusing for users too.

Of course I can adapt to that if the current state was intended.

Comment (ThoWagen) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: And please enjoy your vacation! I can wait until next week for further responses :)

Comment (EreMaijala) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: It's intentional, but I'm open to revising it if it's causing problems. The intention was just to improve consistency (and fix some quirks caused by the inconsistency), because we had two different paths depending on whether the exclude option was displayed.

Comment (EreMaijala) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: I suppose I could change it once more and separate the badge and exclude link from each other.

Comment (demiankatz) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: I see what @ThoWagen means about the non-clickable badges. That does feel a little wrong to me as well. I'm not sure what the best solution is; it might be structurally simpler to just put a second link tag around the badge, but that likely would have bad side effects for screen readers.

Comment (EreMaijala) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: Well, since the badge was part of the link before in case exclude wasn't enabled, it shouldn't be a major issue. I'll check that it doesn't cause any new issues and post a PR if things work out.

Comment (ThoWagen) [reply themes/bootstrap3/templates/Recommend/SideFacets/single-facet.phtml:63]: Thanks @EreMaijala and @demiankatz!

**Pull request merged**
